### PR TITLE
Allow redirects to whitelisted hosts with ports

### LINKED ISSUE #279: Whitelist domain not working for Host:Port on redirects
I have a prometheus instance running under `metrics-qa.staging.com:9043`. I have to give the port number as haproxy on that host uses 9043 and cannot run on any default port of 443.

I have oauth2-proxy v3.2.0 deployed in K8s and exposed under `oauth2-qa.staging.arm.com:9043`. It has below configs set:

      --http-address=0.0.0.0:4180
      --cookie-domain=.staging.com
      --oidc-issuer-url=https://dex-qa.staging.com:9043
      --provider=oidc
      --scope=openid profile email groups
      --ssl-insecure-skip-verify=true
      --whitelist-domain=.staging.com
      --config=/etc/oauth2_proxy/oauth2_proxy.cfg

Here are annotation of prometheus ingress :

    nginx.ingress.kubernetes.io/auth-signin: https://oauth2-qa.staging.com:9043/oauth2/start?rd=$scheme://$best_http_host$request_uri
    nginx.ingress.kubernetes.io/auth-url: https://oauth2-qa.staging.com:9043/oauth2/auth
    nginx.ingress.kubernetes.io/backend-protocol: HTTPS

On successful authentication this redirects to oauth2-qa.staging.com:9043/ and results in 404

## Expected Behavior

It should redirect to https://metrics-qa.staging.com:9043/

On console of chrome I could see the url is formed correctly but redirect still doesnt happen ie

` Request URL: https://oauth2-qa.staging.com:9043/oauth2/start?rd=https://metrics-qa.staging.com:9043/ `

- Version used:
v3.2.0

### PR DESCRIPTION
**Pull request by kamaln7**
This updates `IsValidRedirect` to allow redirects to whitelisted hosts if a port is set on the redirect URL. Fixes #279.

## Description

Use `URL.Hostname()` instead of `URL.Host` to strip the port when validating the host against the list of whitelisted domain.

## How Has This Been Tested?

Added tests to cover the new use cases. I wasn't sure if I needed to add tests for _invalid_ hosts with ports but let me know and I can add that in.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
@kamaln7 Thanks for submitting this PR, I've got an idea about potential improvements, I wonder if we could define some logic so that the whitelist domains could include the port? If no port is set we assume there should be no port set on the request for redirection (ie it's a standard port), if a port is set on the redirect then it must match a whitelisted domain/port combination, WDYT?

> Added tests to cover the new use cases. I wasn't sure if I needed to add tests for invalid hosts with ports but let me know and I can add that in.

If you could add some invalid host tests that would be great!

Finally, please make sure to update the changelog and documentation before this gets merged  🙂

**Comment by kamaln7** [comment]
Thanks for the feedback! I'll be sure to update the documentation and changelog, but I want to confirm that the current behavior is what you were looking for.

Provided that the redirect URL matches a whitelisted domain:

* If the whitelisted domain doesn't contain a port, the redirect URL is considered valid regardless of its port
* If the whitelisted domain contains a port, the redirect URL will be considered valid only if it matches the same port

I've added some more tests which I believe cover all cases but an extra set of eyes is never a bad idea.

Now, there's something I'm not 100% sure about. Because the whitelisted domains aren't proper URLs, only hosts, I don't use `url.Parse()` but rather create a `url.URL` struct manually and set the `Host` field myself. Then I use `URL.Hostname()` and `URL.Port()` as with the redirect URLs.

The code for `URL.Hostname()` and `URL.Port()` doesn't use any fields other than `Host`, so it's fine and it works. But my concern is that by creating the struct manually and not using `url.Parse()`, the `Host` isn't validated. That should be fine if you trust operators to use proper domains in the config, but I feel like that might cause unpredictable behavior if an operator uses an invalid host. I do check if `URL.Hostname()` returns an empty string though I don't know if that's good enough. 

I'm not sure how deep you want to go into this so I'd like to hear your thoughts.

**Comment by JoelSpeed** [comment]
> If the whitelisted domain doesn't contain a port, the redirect URL is considered valid regardless of its port
If the whitelisted domain contains a port, the redirect URL will be considered valid only if it matches the same port

I think if there is no port set on the whitelist domain then we should assume that the request must be on the default port for the protocol (so 80 for http, 443 for https).

Though I would assume that the request wouldn't have a port set when you form it if it's to the default port so I think the check would be to check for an empty port, might be worth testing that assumption though 🤔

**Comment by kamaln7** [comment]
Yes, good point about the port being empty.

I’ll update the code. For wildcard ports I guess we could use “*” as the port assuming that wouldn’t cause issues with parsing. That way it would be backwards compatible too.

**Comment by JoelSpeed** [comment]
That sounds like a really good approach! Please make sure there are sufficient tests for all the cases when you make these changes

**Comment by kamaln7** [comment]
Let me know what you think. I had to copy `splitHostPort` and `validOptionalPort` from `net/url` to the package and modify `validOptionalPort` to allow `*` as a port. Neither `net/url.URL.splitHostPort()` nor `net.SplitHostPort()` work without workarounds.

**Comment by kamaln7** [comment]
Hey @JoelSpeed any updates? 😄

File: oauthproxy.go [hunk 1]
```diff
@@ -504,11 +541,28 @@ func (p *OAuthProxy) IsValidRedirect(redirect string) bool {
 		if err != nil {
 			return false
 		}
+		redirectHostname := redirectURL.Hostname()
+
 		for _, domain := range p.whitelistDomains {
-			if (redirectURL.Host == domain) || (strings.HasPrefix(domain, ".") && strings.HasSuffix(redirectURL.Host, domain)) {
-				return true
+			domainHostname, domainPort := splitHostPort(strings.TrimLeft(domain, "."))
+			if err != nil || domainHostname == "" {
```
Comment (JoelSpeed) [inline oauthproxy.go]: We have no `err` around this part of the code, doesn't seem to make sense to check it here right?

Comment (kamaln7) [reply oauthproxy.go]: You're right, this seems to be left over from old code. Good catch

**Comment by kamaln7** [comment]
No worries, thanks for reviewing!

**Comment by mnaser** [comment]
This is super useful and it seems to have been forgotten till it hit merge conflict.

Perhaps @kamaln7 could be kind enough to rebase and we can help land this? :)

**Comment by syscll** [comment]
Once conflicts are fixed we can merge.

**Comment by kamaln7** [comment]
Fixed! 😃

**Comment by mnaser** [comment]
@syscll can this be merged now? :)

**Comment by syscll** [comment]
Can you give this a final review @JoelSpeed @steakunderscore

**Pull request merged**